### PR TITLE
Possibility to apply default values for the processing gui from the s…

### PR DESCRIPTION
**Pull request by mhugent**
This PR adds the possibility to configure GUI default values for processing in the QGIS.ini file. Use case is that sometimes we want to configure organization wide default values for some algorithms (e.g. precision (GRID_SIZE) for vector overlay). In such cases, it would be handy if the admin could configure it in the prepared ini-file for all users. If the users then open the toolbox to execute an algorithm, the default values will be there. To set the default value e.g. for GRID_SIZE for the intersection algorithm, the admin could write the following into the ini-file:

[Processing]
DefaultGuiParam\native%3Aintersection\GRID_SIZE=0.01

Do you think this is a usefull addition or is the use-case to special (or is there a better way to set default gui values that I did not see)? In case it is usefull, it would probably also be worth to have a gui in the options dialog to add default values.

### CODE REVIEW

File: src/core/processing/qgsprocessingparameters.cpp [hunk 1]
```diff
@@ -2451,6 +2452,40 @@ QgsProcessingParameterDefinition::QgsProcessingParameterDefinition( const QStrin
...
+  return mGuiDefault.isValid() ? mGuiDefault : mDefault;
+}
+
+QVariant QgsProcessingParameterDefinition::defaultGuiValueFromSetting() const
+{
+  if ( mAlgorithm )
+  {
+    QgsSettings s;
+    QVariant settingValue = s.value( QStringLiteral( "/Processing/DefaultGuiParam/%1/%2" ).arg( mAlgorithm->id() ).arg( mName ) );
+    if ( !settingValue.isNull() )
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.cpp:2481]:
```suggestion
    if ( settingValue.isValid() )
```

File: src/core/processing/qgsprocessingparameters.cpp [hunk 2]
```diff
@@ -2451,6 +2452,40 @@ QgsProcessingParameterDefinition::QgsProcessingParameterDefinition( const QStrin
...
+  {
+    return defaultSettingsValue;
+  }
+  return mGuiDefault;
+}
+
+QVariant QgsProcessingParameterDefinition::defaultValueForGui() const
+{
+  QVariant defaultSettingsValue = defaultGuiValueFromSetting();
+  if ( !defaultSettingsValue.isNull() )
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.cpp:2468]:
```suggestion
  if ( defaultSettingsValue.isValid() )
```

File: src/core/processing/qgsprocessingparameters.cpp [hunk 3]
```diff
@@ -2451,6 +2452,40 @@ QgsProcessingParameterDefinition::QgsProcessingParameterDefinition( const QStrin
   , mFlags( optional ? FlagOptional : 0 )
 {}
 
+QVariant QgsProcessingParameterDefinition::guiDefaultValueOverride() const
+{
+  QVariant defaultSettingsValue = defaultGuiValueFromSetting();
+  if ( !defaultSettingsValue.isNull() )
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.cpp:2458]:
```suggestion
  if ( defaultSettingsValue.isValid() )
```

File: src/core/processing/qgsprocessingparameters.h [hunk 4]
```diff
@@ -896,6 +896,13 @@ class CORE_EXPORT QgsProcessingParameterDefinition
     QVariant valueAsJsonObjectPrivate( const QVariant &value, QgsProcessingContext &context, ValueAsStringFlags flags ) const;
 #endif
 
+    /**
+     * Default gui value for an algorithm parameter from settings
+     *
+     * \return default value from settings or null QVariant if there is no default value defined in the settings
+     */
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.h:903]: \since QGIS 3.34

File: src/core/processing/qgsprocessingparameters.h [hunk 5]
```diff
@@ -896,6 +896,13 @@ class CORE_EXPORT QgsProcessingParameterDefinition
     QVariant valueAsJsonObjectPrivate( const QVariant &value, QgsProcessingContext &context, ValueAsStringFlags flags ) const;
 #endif
 
+    /**
+     * Default gui value for an algorithm parameter from settings
+     *
+     * \return default value from settings or null QVariant if there is no default value defined in the settings
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.h:902]:
```suggestion
     * \return default value from settings or invalid QVariant if there is no default value defined in the settings
```

**Comment by nyalldawson** [comment]
I think this is a great idea -- I can definitely see the use case in organisation (or user) specific overrides for default values. :+1: 

Can you add tests covering this too?

**Comment by mhugent** [comment]
Thanks for reviewing.
I've added a unit test to the PR.

**Review by nyalldawson — APPROVED** [review]

**Comment by DelazJ** [comment]
> In case it is usefull, it would probably also be worth to have a gui in the options dialog to add default values.

@mhugent To be sure, the above suggestion has never been implemented, right? The only way to add default values for algs currently is by editing the ini file. Am I right?

**Pull request merged**